enigma-swing/src/main/java/org/quiltmc/enigma/gui/panel/EntryTooltip.java

```diff
@@ -0,0 +1,845 @@
+			}
+		});
+
+		this.addMouseMotionListener(new MouseAdapter() {
+			@Override
+			public void mouseDragged(MouseEvent e) {
+				final Point dragStart = EntryTooltip.this.dragStart;
+				if (dragStart != null) {
+					final Point pos = EntryTooltip.this.getLocation();
+					pos.translate(e.getX() - dragStart.x, e.getY() - dragStart.y);
+					EntryTooltip.this.setLocation(pos);
+				}
+			}
+		});
+
+		this.addWindowFocusListener(new WindowAdapter() {
+			@Override
+			public void windowLostFocus(WindowEvent e) {
+				EntryTooltip.this.close();
+			}
+		});
+	}
+
+	// setting/overriding max size does not work; getMaximumSize is never called
+	@Override
+	public Dimension getPreferredSize() {
+		final Dimension superPreferred = super.getPreferredSize();
+		final Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
+		return new Dimension(
+			Math.min(Math.min(superPreferred.width, screenSize.width / 2), ScaleUtil.scale(DEFAULT_MAX_WIDTH)),
+			Math.min(superPreferred.height, screenSize.height / 2)
+		);
+	}
+
+	// Sometimes when re-populating and resizing+moving, the cursor may be briefly over the parent EditorPanel.
+	// This is used to stop EditorPanel from starting its mouseStoppedMovingTimer which may reset the tooltip to the
+	// token under the cursor, discarding the re-populated content.
+	public boolean hasRepopulated() {
+		return this.repopulated;
+	}
+
+	/**
+	 * Opens this tooltip and populates it with information about the passed {@code target}.
+	 *
+	 * @param target        the entry whose information will be displayed
+	 * @param inherited     whether this tooltip is displaying information about the parent of another entry
+	 * @param eventReceiver a component to receive events such as key presses that cause this tooltip to close;
+	 *                      may be {@code null}
+	 */
+	public void open(Entry<?> target, boolean inherited, @Nullable Component eventReceiver) {
+		this.eventReceiver = eventReceiver;
+		this.populateWith(target, inherited, true);
+		this.setVisible(true);
+
+		this.addExternalListeners();
+	}
+
+	private void populateWith(Entry<?> target, boolean inherited, boolean opening) {
+		this.repopulated = !opening;
+		this.content.removeAll();
+
+		@Nullable
+		final MouseAdapter stopInteraction = Config.editor().entryTooltips.interactable.value()
+				? null : new MouseAdapter() {
+					@Override
+					public void mousePressed(MouseEvent e) {
+						// dispatching mouse events here causes cast exceptions when
+						// receivers get unexpected event sources
+						EntryTooltip.this.close();
+						e.consume();
+					}
+				};
+
+		final Font editorFont = ScaleUtil.scaleFont(Config.currentFonts().editor.value());
+		final Font italEditorFont = ScaleUtil.scaleFont(Config.currentFonts().editor.value().deriveFont(Font.ITALIC));
+
+		int gridY = 0;
+
+		{
+			final Box parentLabelRow = Box.createHorizontalBox();
+
+			final JLabel from = labelOf(inherited ? "inherited from" : "from", italEditorFont);
```

Comment:
would be nice if this was translatable! doesn't need retranslation logic, but make sure it uses formatting (%s) to account for different semantics

Reply:
i agree, it should!